drivers/sensor/bq274xx/bq274xx.c

```diff
@@ -0,0 +1,584 @@
+	int status = 0;
+	u8_t tmp_checksum = 0, checksum_old = 0, checksum_new = 0;
+	u16_t flags = 0, designenergy_mwh = 0, taperrate = 0, id;
+	u8_t designcap_msb, designcap_lsb, designenergy_msb, designenergy_lsb,
+	     terminatevolt_msb, terminatevolt_lsb, taperrate_msb, taperrate_lsb;
+	u8_t block[32];
+
+	designenergy_mwh = (u16_t) 3.7 * DT_INST_0_TI_BQ274XX_DESIGN_CAPACITY;
+	taperrate = (u16_t) DT_INST_0_TI_BQ274XX_DESIGN_CAPACITY / (0.1 * DT_INST_0_TI_BQ274XX_TAPER_CURRENT);
+
+	bq274xx->i2c = device_get_binding(DT_INST_0_TI_BQ274XX_BUS_NAME);
+	if (bq274xx == NULL) {
+		LOG_ERR("Could not get pointer to %s device.",
+			  DT_INST_0_TI_BQ274XX_BUS_NAME);
+		return -EINVAL;
+	}
+
+	status = bq274xx_get_device_type(bq274xx, &id);
+	if (status < 0) {
+		LOG_ERR("Unable to get device ID");
+		return -EIO;
+	}
+
+	if (id != BQ274XX_DEVICE_ID) {
+		LOG_ERR("Invalid Device");
+		return -EINVAL;
+	}
+
+	/** Unseal the battery control register **/
+	status = bq274xx_control_reg_write(bq274xx, BQ274XX_UNSEAL_KEY);
+	if (status < 0) {
+		LOG_ERR("Unable to unseal the battery");
+		return -EIO;
+	}
+
+	status = bq274xx_control_reg_write(bq274xx, BQ274XX_UNSEAL_KEY);
+	if (status < 0) {
+		LOG_ERR("Unable to unseal the battery");
+		return -EIO;
+	}
+
+	/* Send CFG_UPDATE */
+	status = bq274xx_control_reg_write(bq274xx, BQ274XX_CONTROL_SET_CFGUPDATE);
+	if (status < 0) {
+		LOG_ERR("Unable to set CFGUpdate");
+		return -EIO;
+	}
+
+	/** Step to place the Gauge into CONFIG UPDATE Mode **/
+	do {
+		status = bq274xx_command_reg_read(bq274xx, BQ274XX_COMMAND_FLAGS, &flags);
+		if (status < 0) {
+			LOG_ERR("Unable to read flags");
+			return -EIO;
+		}
+
+		if (!(flags & 0x0010)) {
+			k_sleep(BQ274XX_SUBCLASS_DELAY * 10);
+		}
+
+	} while (!(flags & 0x0010));
+
+	status = bq274xx_command_reg_write(bq274xx, BQ274XX_EXTENDED_DATA_CONTROL, 0x00);
+	if (status < 0) {
+		LOG_ERR("Failed to enable block data memory");
+		return -EIO;
+	}
+
+	/* Access State subclass */
+	status = bq274xx_command_reg_write(bq274xx, BQ274XX_EXTENDED_DATA_CLASS, 0x52);
+	if (status < 0) {
+		LOG_ERR("Failed to update state subclass");
+		return -EIO;
+	}
+
+	/* Write the block offset */
+	status = bq274xx_command_reg_write(bq274xx, BQ274XX_EXTENDED_DATA_BLOCK, 0x00);
+	if (status < 0) {
+		LOG_ERR("Failed to update block offset");
+		return -EIO;
+	}
+
+	for (u8_t i = 0; i < 32; i++) {
+		block[i] = 0;
+	}
+
+	status = bq274xx_read_data_block(bq274xx, 0x00, block, 32);
+	if (status < 0) {
+		LOG_ERR("Unable to read block data");
+		return -EIO;
+	}
+
+	tmp_checksum = 0;
+	for (u8_t i = 0; i < 32; i++) {
+		tmp_checksum += block[i];
+	}
+	tmp_checksum = 255 - tmp_checksum;
+
+	/* Read the block checksum */
+	status = i2c_reg_read_byte(bq274xx->i2c, BQ274XX_I2C_ADDRESS,
+				BQ274XX_EXTENDED_CHECKSUM, &checksum_old);
+	if (status < 0) {
+		LOG_ERR("Unable to read block checksum");
+		return -EIO;
+	}
+
+	designcap_msb = DT_INST_0_TI_BQ274XX_DESIGN_CAPACITY >> 8;
```

Comment:
In case you plan to support more than a single instance, the `cfg_info` from `DEVICE_AND_API_INIT` field can be used to store constant configuration. This way driver code does not depend on the instance number.

Reply:
@gmarull I understand moving the static DT properties to cfg_info makes sense. But I don't understand "support more than a single instance" for I2C device?

Reply:
Current driver code can only handle a single sensor instance. In order to account for any number of instances it is better to decouple driver code from DT constants. One way to do that is to store DT constants into the `static const` configuration structure and then provide a custom initializer, e.g.

```c
#define MYSENSOR_INIT(index)                                        \
                                                                    \
    static const struct mysensor_config mysensor_config_##index = { \
	    .prop_name = DT_INST_PROP(index, prop_name)                 \
    };                                                              \
                                                                    \
    static struct mysensor_data mysensor_data_##index = {           \
	    .other_prop_name = DT_INST_PROP(index, other_prop_name)     \
    };                                                              \
                                                                    \
    DEVICE_AND_API_INIT(mysensor_##index, DT_INST_LABEL(index),     \
		    &mysensor_init,                                         \
		    &mysensor_data_##index, &mysensor_cfg_##index,          \
		    POST_KERNEL, CONFIG_SENSOR_INIT_PRIORITY,               \
		    &mysensor_api);                                         \
									                                 

DT_INST_FOREACH(MYSENSOR_INIT)
```

Reply:
@gmarull Thanks for the details. Fixed.